File: bnhsdk/android/src/main/kotlin/ai/binah/binah_flutter_sdk/SessionManager.kt
package ai.binah.binah_flutter_sdk

import ai.binah.sdk.api.HealthMonitorException
import ai.binah.sdk.api.SessionEnabledVitalSigns
import ai.binah.sdk.api.images.DeviceOrientation
import ai.binah.sdk.api.images.ImageData
import ai.binah.sdk.api.images.ImageListener
import ai.binah.sdk.api.license.LicenseDetails
import ai.binah.sdk.api.license.LicenseInfo
import ai.binah.sdk.api.session.Session
import ai.binah.sdk.api.session.SessionInfoListener
import ai.binah.sdk.api.session.SessionState
import ai.binah.sdk.api.alerts.ErrorData
import ai.binah.sdk.api.alerts.WarningData
import ai.binah.sdk.api.session.demographics.Sex
import ai.binah.sdk.api.session.demographics.SubjectDemographic
import ai.binah.sdk.api.vital_signs.VitalSign
import ai.binah.sdk.api.vital_signs.VitalSignsListener
import ai.binah.sdk.api.vital_signs.VitalSignsResults
import ai.binah.sdk.session.FaceSessionBuilder
import ai.binah.sdk.session.FingerSessionBuilder
import ai.binah.sdk.session.MeasurementMode
import android.content.Context
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow

object SessionManager: ImageDataSource, ImageListener, VitalSignsListener, SessionInfoListener {

    private val _images = MutableSharedFlow<ImageData>(replay = 0, extraBufferCapacity = 1, onBufferOverflow = BufferOverflow.DROP_LATEST)
    private var session: Session? = null

    override val images: Flow<ImageData> = _images

    var eventChannel: BinahEventChannel? = null

    @Throws(HealthMonitorException::class)
    fun createSession(
        context: Context,
        licenseKey: String,
        productId: String? = null,
        measurementMode: Int?,
        deviceOrientation: Int? = null,
        subjectSex: Int? = null,
        subjectAge: Double? = null,
        subjectWeight: Double? = null,
        detectionAlwaysOn: Boolean? = false,
        options: Map<String, Any>? = null
    ) {
        val mode = resolveMeasurementMode(measurementMode)
        val sessionBuilder = if (mode == MeasurementMode.FACE) {
            val orientation = resolveDeviceOrientation(deviceOrientation)
            val subjectDemographic = resolveSubjectDemographic(
                subjectSex,
                subjectAge,
                subjectWeight
            )

            FaceSessionBuilder(context)
                .withDeviceOrientation(orientation)
                .withSubjectDemographic(subjectDemographic)
                .withDetectionAlwaysOn(detectionAlwaysOn ?: false)
        } else {
            FingerSessionBuilder(context)
        }

        session = sessionBuilder
            .withImageListener(this@SessionManager)
            .withVitalSignsListener(this@SessionManager)
            .withSessionInfoListener(this@SessionManager)
            .withOptions(options)
            .build(LicenseDetails(licenseKey, productId))
    }

    @Throws(HealthMonitorException::class)
    fun startSession(duration: Int?) {
        session?.start(duration?.toLong() ?: 0)
    }

    @Throws(HealthMonitorException::class)
    fun stopSession() {
        session?.stop()
    }

    fun terminateSession() {
        session?.terminate()
    }

    fun getSessionState(): SessionState? {
        return session?.state
    }

    override fun onVitalSign(vitalSign: VitalSign) {
        eventChannel?.sendEvent(NativeBridgeEvents.sessionVitalSign, vitalSign.toMap() ?: return)
    }

    override fun onFinalResults(vitalSignsResults: VitalSignsResults) {
        val results = vitalSignsResults.results.mapNotNull { result ->
            result.toMap()
        }
        eventChannel?.sendEvent(NativeBridgeEvents.sessionFinalResults, results)
    }

    override fun onImage(imageData: ImageData) {
        eventChannel?.sendEvent(NativeBridgeEvents.imageData, imageData.toMap())
        _images.tryEmit(imageData)
    }

    override fun onSessionStateChange(sessionState: SessionState) {
        eventChannel?.sendEvent(NativeBridgeEvents.sessionStateChange, sessionState.ordinal)
    }

    override fun onWarning(warningData: WarningData) {
        eventChannel?.sendEvent(NativeBridgeEvents.sessionWarning, warningData.toMap())
    }

    override fun onError(errorData: ErrorData) {
        eventChannel?.sendEvent(NativeBridgeEvents.sessionError, errorData.toMap())
    }

    override fun onLicenseInfo(licenseInfo: LicenseInfo) {
        eventChannel?.sendEvent(NativeBridgeEvents.licenseInfo, licenseInfo.toMap())
    }

    override fun onEnabledVitalSigns(enabledVitalSigns: SessionEnabledVitalSigns) {
        eventChannel?.sendEvent(NativeBridgeEvents.enabledVitalSigns, enabledVitalSigns.toMap())
    }

    private fun resolveMeasurementMode(measurementMode: Int?): MeasurementMode {
        if (measurementMode == MeasurementMode.FINGER.ordinal) {
            return MeasurementMode.FINGER
        }

        return MeasurementMode.FACE
    }

    private fun resolveDeviceOrientation(deviceOrientation: Int?): DeviceOrientation? {
        return deviceOrientation?.let {orientation ->
            try {
                DeviceOrientation.values()[orientation]
            } catch (ignore: IndexOutOfBoundsException) {
                null
            }
        }
    }

    private fun resolveSubjectDemographic(sexInt: Int?, age: Double?, weight: Double?): SubjectDemographic {
        val sex = sexInt?.let { sex ->
            try {
                Sex.values()[sex]
            } catch (e: IndexOutOfBoundsException) {
                Sex.UNSPECIFIED
            }
        }

        return SubjectDemographic(sex, age, weight)
    }
}

File: bnhsdk/ios/Classes/SessionManager.swift

import Foundation
import Combine

class SessionManager: ImageDataSource, ImageListener, VitalSignsListener, SessionInfoListener {
    
    static let shared = SessionManager()
    
    var eventChannel: BinahEventChannel?
        private var session: Session?
    
    let images = PassthroughSubject<ImageData, Never>()
    
    private init() {}
    
    func createSession(
        licenseKey: String,
        productId: String? = nil,
        measurementMode: Int?,
        deviceOrientation: Int? = nil,
        subjectSex: Int? = nil,
        subjectAge: Double? = nil,
        subjectWeight: Double? = nil,
        detectionAlwaysOn: Bool? = false,
        imageFormatMode: Int? = nil,
        options: [String: Any]? = nil
    
    ) throws {
        let mode = resolveMeasurementMode(measurementMode: measurementMode)
        var sessionBuilder: SessionBuilder
        if (mode == MeasurementMode.face) {
            var builder = FaceSessionBuilder()
            if let subjectDemographic = resolveSubjectDemographic(sex: subjectSex, age: subjectAge, weight: subjectWeight) {
                builder = builder.withSubjectDemographic(subjectDemographic)
            }
            
            if let orientation = resolveDeviceOrientation(deviceOrientation: deviceOrientation) {
                builder = builder.withDeviceOrientation(orientation)
            }

            if let imageFormat = resolveImageFormat(imageFormatMode: imageFormatMode) {
                builder = builder.withImageFormatMode(imageFormat)
            }

            builder = builder.withDetectionAlwaysOn(detectionAlwaysOn ?? false)
            sessionBuilder = builder
        } else {
            sessionBuilder = FingerSessionBuilder()
        }

        session = try sessionBuilder
            .withImageListener(self)
            .withVitalSignsListener(self)
            .withSessionInfoListener(self)
            .withOptions(options: options ?? [:])
            .build(licenseDetails: LicenseDetails(licenseKey: licenseKey, productId: productId))
    }
    
    
    func startSession(duration: Int?) throws {
        try session?.start(measurementDuration: UInt64(duration ?? 0))
    }

    func stopSession() throws {
        try session?.stop()
    }

    func terminateSession() {
        session?.terminate()
    }

    func getSessionState() -> SessionState? {
        return session?.state
    }

    func onImage(imageData: ImageData) {
        images.send(imageData)
        eventChannel?.sendEvent(name: NativeBridgeEvents.imageData , payload: imageData.toMap())
    }
    
    func onVitalSign(vitalSign: VitalSign) {
        if let map = vitalSign.toMap() {
            eventChannel?.sendEvent(name: NativeBridgeEvents.sessionVitalSign , payload: map)
        }
    }
    
    func onFinalResults(results: VitalSignsResults) {
        let finalResults = results.getResults().compactMap { result in
            result.toMap()
        }
        
        eventChannel?.sendEvent(name: NativeBridgeEvents.sessionFinalResults, payload: finalResults)
    }
    
    func onSessionStateChange(sessionState: SessionState) {
        eventChannel?.sendEvent(name: NativeBridgeEvents.sessionStateChange, payload: sessionState.rawValue)
    }
    
    func onWarning(warningData: WarningData) {
        eventChannel?.sendEvent(name: NativeBridgeEvents.sessionWarning, payload: warningData.toMap())
    }
    
    func onError(errorData: ErrorData) {
        eventChannel?.sendEvent(name: NativeBridgeEvents.sessionError, payload: errorData.toMap())
    }
    
    func onLicenseInfo(licenseInfo: LicenseInfo) {
        eventChannel?.sendEvent(name: NativeBridgeEvents.licenseInfo, payload: licenseInfo.toMap())
    }
    
    func onEnabledVitalSigns(enabledVitalSigns: SessionEnabledVitalSigns) {
        eventChannel?.sendEvent(name: NativeBridgeEvents.enabledVitalSigns, payload: enabledVitalSigns.toMap())
    }

    private func resolveMeasurementMode(measurementMode: Int?) -> MeasurementMode {
        if (measurementMode == MeasurementMode.finger.rawValue) {
            return MeasurementMode.finger
        }

        return MeasurementMode.face
    }

    private func resolveDeviceOrientation(deviceOrientation: Int?) -> DeviceOrientation? {
        guard let orientation = deviceOrientation, let orientationEnum = DeviceOrientation.init(rawValue: orientation) else {
            return nil
        }
        
        return orientationEnum
    }

    private func resolveSubjectDemographic(sex: Int?, age: Double?, weight: Double?) -> SubjectDemographic? {
        if (sex == nil && age == nil && weight == nil) {
            return nil
        }
        
        var sdkAge: NSNumber?
        if let age = age {
            sdkAge = NSNumber.init(value: age)
        }
        
        var sdkWeight: NSNumber?
        if let weight = weight {
            sdkWeight = NSNumber.init(value: weight)
        }
        
        return SubjectDemographic(
            sex: Sex.init(rawValue: sex ?? 0) ?? Sex.unspecified,
            age: sdkAge,
            weight: sdkWeight)
    }

    private func resolveImageFormat(imageFormatMode: Int?) -> ImageFormatMode? {
        guard let imageFormat = imageFormatMode,
              let imageFormatEnum = ImageFormatMode.init(rawValue: imageFormat) else {
            return nil
        }

        return imageFormatEnum
    }
}
